### DISCUSSION TITLE
Image storage options

### DISCUSSION THREAD

**Discussion by bglamadrid** [trebol-ecommerce/trebol-backend-monolith#156] [Ideas]
It is simply a bad practice to store image uploads in databases, let alone passing their raw data through memory. This is why the `Image` entity class contains `String url` and `String filename` fields, but not something like `byte[] rawData`.

If you ask me, I would recur to a SaaS solution that offers the ability to upload and deliver this content. But of course, there could cheaper solutions, like self-hosting a file server.

Regardless of the opted solutions, there needs to be integration code in place for anything to be used. And most importantly, there should be a simple Java API that makes it easy to switch implementations.

**Comment by mutasDev** [comment]
Would it be fine to make a storageservice that stores the images on the hard disk, with a reference UUID and store this UUID (resolves to the path of image on the disk) in the db of products?

**Comment by bglamadrid** [comment]
> Would it be fine to make a storageservice that stores the images on the hard disk, with a reference UUID and store this UUID (resolves to the path of image on the disk) in the db of products?

Yes, it could work. 

Still I'd like to separate its implementation logic, abstracting to an interface that enables to use different, maybe multiple storage solutions at once.

Problem is, `Image` entities and related API requests have little metadata to support anything but sending hot links to already stored images. It is not yet designed to support the uploading process itself.

I think I should work a little more on the API before, to make both ends comply.
